### ISSUE TITLE
Server does not verify nounces

### ISSUE THREAD

**Issue by krestenkrab** [jlouis/curve_tun#1]
CurveZMQ requires that the server validate that message nounces are increasing by one for each individual message. Perhaps it would be a good idea to do the same here.  I pressume that this guards agains replay attack, as well as ensures against lost messages; it also ensures that client that might expose their keys indirectly by reusing nounces would be declined connectivity.

I'd suggest that some initial counter is established in the initial exchange, and then going forward these are required to be increasing by 1 for each successive message, perhaps modulus the counter size so that the initial value can be any random number in the range.

**Comment by jlouis** [comment]
I've implemented a fix for this in 1cc5c4d2e2912e4ef47c7a91d551e5fb2a149b99 and on to a63a9b5 where the counter is enforced to be increasing by one every message. This guards against replay attacks, and does some provisioning against implementation failures as you say. The counter is established once we enter the connected state as the state key `rc`, and is verified in each subsequent message. Nonce uniqueness is achieved because the counter operates on a short-term key (and can thus start from 0), and the server side has a different nonce from the client side, because text is appended to the nonce message with the identifying side.

While here, I've also rewritten the code a bit and have attached an initial Common Test case in a suite to make it easy to test for correctness. The code is still a mess, but establishing a baseline through a test case is the right way to handle refactoring later.

**Comment by krestenkrab** [comment]
Looks good to me.

**Comment by krestenkrab** [comment]
On second thought -- and as I'm working on a minimal C implementation for this -- it would be great if these sequence numbers can start from 1 in stead of 0; as that allows me to use an uint64_t for the `rn` value, without special-casing the first message.  Thus this can be initialised to 0 rather than -1 in the code.  Thoughts?

**Comment by jlouis** [comment]
It makes the C code simpler since it doesn't have to battle with signed integers. Which is a good thing. An error is probably more likely to be in the C code than in the Erlang code.

It does allow for a "hole" in the 0-slot but since no client will ever send this, it can't be replayed against. Losing a single packet in the counter isn't a problem in a 64bit unsigned counter. If you ever hit its limit, you are sending more than 500 billion messages per second for a year, or any lower rate for more years.

Long story short: I don't think it is a problem.

**Comment by krestenkrab** [comment]
OK, looks like a sequence like this should solve the problem:

``` c
    // validate sequence number
    rn = read64be(data+8);
    if (rn < (ctx->rn + 1ULL)) { return -1; }
    ctx->rn = rn;
```

both `rn` and `ctx->rn` are `uint64_t`; and if the latter is initialised to `(uint64_t)-1` then it should be just fine AFAIKS.
